akka-stream/src/main/scala/akka/stream/javadsl/Source.scala

```diff
@@ -320,20 +304,124 @@ class Source[+Out, +Mat](delegate: scaladsl.Source[Out, Mat]) extends Graph[Sour
+   *
+   * '''Emits when''' element is available from current source or from the given [[Source]] when current is completed
+   *
+   * '''Backpressures when''' downstream backpressures
+   *
+   * '''Completes when''' given [[Source]] completes
+   *
+   * '''Cancels when''' downstream cancels
+   */
+  def concat[T >: Out, M](that: Graph[SourceShape[T], M]): javadsl.Source[T, Mat] =
+    new Source(delegate.concat(that))
+
+  /**
+   * Concatenate this [[Source]] with the given one, meaning that once current
+   * is exhausted and all result elements have been generated,
+   * the given source elements will be produced.
+   *
+   * Note that given [[Source]] is materialized together with this Flow and just kept
+   * from producing elements by asserting back-pressure until its time comes.
+   *
+   * If this [[Source]] gets upstream error - no elements from the given [[Source]] will be pulled.
+   *
+   * @see [[#concat]].
+   */
+  def concatMat[T >: Out, M, M2](that: Graph[SourceShape[T], M],
+                                 matF: function.Function2[Mat, M, M2]): javadsl.Source[T, M2] =
+    new Source(delegate.concatMat(that)(combinerToScala(matF)))
+
+  /**
+   * Merge the given [[Source]] to the current one, taking elements as they arrive from input streams,
+   * picking randomly when several elements ready.
+   *
+   * '''Emits when''' one of the inputs has an element available
+   *
+   * '''Backpressures when''' downstream backpressures
+   *
+   * '''Completes when''' all upstreams complete
+   *
+   * '''Cancels when''' downstream cancels
+   */
+  def merge[T >: Out](that: Graph[SourceShape[T], _]): javadsl.Source[T, Mat] =
+    new Source(delegate.merge(that))
+
+  /**
+   * Merge the given [[Source]] to the current one, taking elements as they arrive from input streams,
+   * picking randomly when several elements ready.
+   *
+   * @see [[#merge]].
    */
-  def concat[Out2 >: Out, M2](second: Graph[SourceShape[Out2], M2]): javadsl.Source[Out2, (Mat, M2)] =
-    Source.concat(this, second)
+  def mergeMat[T >: Out, M, M2](that: Graph[SourceShape[T], M],
+                                matF: function.Function2[Mat, M, M2]): javadsl.Source[T, M2] =
+    new Source(delegate.mergeMat(that)(combinerToScala(matF)))
 
   /**
-   * Concatenates a second source so that the first element
-   * emitted by that source is emitted after the last element of this
-   * source.
+   * Combine the elements of current [[Source]] and the given one into a stream of tuples.
+   *
+   * '''Emits when''' all of the inputs has an element available
+   *
+   * '''Backpressures when''' downstream backpressures
+   *
+   * '''Completes when''' any upstream completes
+   *
+   * '''Cancels when''' downstream cancels
+   */
+  def zip[T](that: Graph[SourceShape[T], _]): javadsl.Source[Out @uncheckedVariance Pair T, Mat] =
+    zipMat(that, combinerToJava((a: Mat, b: Any) ⇒ a))
```

Comment:
Np! Afaik there's a javadsl Keep.left you can pass in insteadof
combinerToJava

-- 
Cheers,
√
On 4 Oct 2015 14:42, "Alexander Golubev" <notifications@github.com> wrote:

> In akka-stream/src/main/scala/akka/stream/javadsl/Source.scala
> <https://github.com/akka/akka/pull/18536#discussion_r41097336>:
>
> >
> >    /**
> > -   * Concatenates a second source so that the first element
> > -   * emitted by that source is emitted after the last element of this
> > -   * source.
> > +   * Combine the elements of current [[Source]] and the given one into a stream of tuples.
> > +   *
> > +   * '''Emits when''' all of the inputs has an element available
> > +   *
> > +   * '''Backpressures when''' downstream backpressures
> > +   *
> > +   * '''Completes when''' any upstream completes
> > +   *
> > +   * '''Cancels when''' downstream cancels
> > +   */
> > +  def zip[T](that: Graph[SourceShape[T], _]): javadsl.Source[Out @uncheckedVariance Pair T, Mat] =
> > +    zipMat(that, combinerToJava((a: Mat, b: Any) ⇒ a))
>
> Ahh, missed that :(
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/akka/akka/pull/18536/files#r41097336>.
>